teps/0004-tekton-integrations-annotations-statuses.md

```diff
@@ -0,0 +1,611 @@
+agnostic to the source of configuration.
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal and how do we mitigate. Think broadly.
+For example, consider both security and how this will impact the larger
+kubernetes ecosystem.
+
+How will security be reviewed and by whom?
+
+How will UX be reviewed and by whom?
+
+Consider including folks that also work outside the WGs or subproject.
+-->
+
+#### External Integration Quotas
+
+If relying on reconcilers / events that trigger on every update, integration
+providers will need to be careful to stay within quotas external providers may
+impose.
+
+While Tekton Pipelines will not impose specific throttling for integration
+providers, it will be a general best practice to use Status/Condition data (e.g.
+`ObservedGeneration`, `Conditions`, `LastTransitionTime`) as well as reconciler
+push back with `RequeueAfter` to stay within external quotas.
+
+### User Experience (optional)
+
+<!--
+Consideration about the user experience. Depending on the area of change,
+users may be task and pipeline editors, they may trigger task and pipeline
+runs or they may be responsible for monitoring the execution of runs,
+via CLI, dashboard or a monitoring system.
+
+Consider including folks that also work on CLI and dashboard.
+-->
+
+#### Storing data in Pipelines
+
+Similar to `TaskRunResults` and `PipelineResourceResults`, integration statuses
+are another form of external artifact generated as a result of the Pipeline.
+Because of this, it makes sense to store these statuses as a part of the
+Pipeline/TaskRun status, so that it is easy to find and see this data when
+something goes wrong.
+
+### Performance (optional)
+
+<!--
+Consideration about performance.
+What impact does this change have on the start-up time and execution time
+of task and pipeline runs? What impact does it have on the resource footprint
+of Tekton controllers as well as task and pipeline runs?
+
+Consider which use cases are impacted by this change and what are their
+performance requirements.
+-->
+
+N/A
+
+For pipelines, this should have no impact on runtime performance. Integration
+providers will need to be mindful of the information they store within the
+Pipeline/TaskRun objects to stay within k8s limits, but we are not exposing
+additional risk that did not already exist.
+
+## Design Details
+
+### Integration Status
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable.  This may include API specs (though not always
+required) or even code snippets.  If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+Most of the fields we want are already included in the Knative's
+[Status](https://pkg.go.dev/knative.dev/pkg/apis/duck/v1beta1?tab=doc#Status)
+object, so we can likely take a similar approach here. e.g.:
+
+```go
+type IntegrationStatus struct {
```

Comment:
Why not just a  new `.status.conditions` item with a new `.type`?

```
status:
  conditions:
  - type: Succeeded
     status: True  # it succeeded!
  - type: SlackNotified
     status: False  # ...but we couldn't notify Slack as requested
     reason: AuthProbably
```

Reply:
Even though I'd like to keep this data localized to the Run, I think we should carve out a separate area within the status for integration data. We also cannot use conditions as is to include extra data that might have been received in response to an integration (e.g. webhook receipt, pull request status ids, etc).

Added an example to make this clearer!

Reply:
That makes sense, thanks for the clarifying example.